### DISCUSSION TITLE
Automatically rename key with user entered data

### DISCUSSION THREAD

**Discussion by lsmith212a** [pimcore/pimcore#12713] [Q&A / Support] [answered]
In the sample Customer data I have noticed that when you add a new customer, the key for the data object is renamed to concatenate the Last Name - First Name for that customer. Then if the user updates the customer name, the key name also updates. 

Is it possible to do the same thing in a product class? How can I apply that same logic to another data class?

**Comment by NiklasBr** [comment, marked as answer]
Yes, [with Event Listeners](https://pimcore.com/docs/pimcore/current/Development_Documentation/Extending_Pimcore/Event_API_and_Event_Manager.html). I suggest the `preUpdate` event, something like this:

```php
namespace App\EventListener;
class MyEventListener implements EventSubscriberInterface
{
    public static function getSubscribedEvents(): array
    {
        return [
            DataObjectEvents::PRE_UPDATE => 'myPreUpdate',
        ];
    }

    public function myPreUpdate(ElementEventInterface $elementEvent)
    {
        $object = $elementEvent->getObject();
        if ($object instanceof Product && !empty($object->getMyFieldName())) {
            $newKey = $object->getMyFieldName();
            // Here you should make sure the Key is unique, I think this could work
            $object->setKey($newKey);
            $object->setKey(
                \Pimcore\Model\Element\Service::getUniqueKey($object)
            );
        }
    }
}
```

**Reply by lsmith212a** [reply]
Thank you. This is very helpful. We are trying to determine if we can modify the classes in this way without a licensed version. Do you know if that's allowed?

**Reply by NiklasBr** [reply]
@lsmith212a Are you trying to edit the classes provided by Pimcore? What are you trying to do that cannot be done with this: https://pimcore.com/docs/pimcore/current/Development_Documentation/Extending_Pimcore/index.html
